src/server/server_family.cc

```diff
@@ -1297,31 +1281,28 @@ void ServerFamily::ClientPause(CmdArgList args, ConnectionContext* cntx) {
     return (*cntx)->SendError(err->MakeReply());
   }
 
-  // Pause dispatch commands before updating client puase state, and enable dispatch after updating
-  // pause state. This will unsure that when we after changing the state all running commands will
-  // read the new pause state, and we will not pause client in the middle of a transaction.
-  service_.proactor_pool().Await([](util::ProactorBase* pb) {
-    ServerState& etl = *ServerState::tlocal();
-    etl.SetPauseDispatch(true);
+  // Set global pause state and track commands that are running when the pause state is flipped.
+  // Exlude already paused commands from the busy count.
+  DispatchTracker tracker{GetListeners(), cntx->conn(), true /* ignore paused commands */};
+  service_.proactor_pool().Await([&tracker, pause_state](util::ProactorBase* pb) {
+    // Commands don't suspend before checking the pause state, so
+    // it's impossible to deadlock on waiting for a command that will be paused.
+    tracker.TrackOnThread();
+    ServerState::tlocal()->SetPauseState(pause_state, true);
   });
```

Comment:
Now I use this instead of the pausing-dispatch fix... Should work if we don't suspend before checking pause state